File: src/go/set_cover.go
package algorithms

import "strings"

func filter(src []string) (res []string) {
	for _, s := range src {
		newStr := strings.Join(res, " ")
		if !strings.Contains(newStr, s) {
			res = append(res, s)
		}
	}
	return
}

func intersections(slice1, slice2 []string) (intersection []string) {
	str1 := strings.Join(filter(slice1), " ")
	for _, s := range filter(slice2) {
		if strings.Contains(str1, s) {
			intersection = append(intersection, s)
		}
	}
	return
}

func diff(slice1, slice2 []string) (diff []string) {
	for _, sl1V := range slice1 {
		found := false
		for _, sl2V := range slice2 {
			if sl1V == sl2V {
				found = true
			}
		}
		if !found {
			diff = append(diff, sl1V)
		}
	}
	return
}

func SetCover(data map[string][]string, needed []string) (finalSet []string) {
	neededCopy := make([]string, len(needed))
	copy(neededCopy, needed)

	for len(neededCopy) != 0 {
		bestFound := ""
		bestCoveredSet := []string{}

		for k, v := range data {
			covered := intersections(v, neededCopy)

			if len(covered) > len(bestCoveredSet) {
				bestFound = k
				bestCoveredSet = covered
			}
		}
		neededCopy = diff(neededCopy, bestCoveredSet)
		finalSet = append(finalSet, bestFound)
	}
	return
}

File: src/php/SetCover.php
<?php

namespace GulDmitry\Algorithms;

/**
 * Greedy algorithm
 */
class SetCover
{
    public function findCoverage(array $data, array $needed): array
    {
        $finalSet = [];
        while (!empty($needed)) {
            $bestFound = null;
            $bestCoveredSet = [];

            foreach ($data as $k => $v) {
                // $covered = $needed & $v;
                $covered = array_intersect($needed, $v);

                if (count($covered) > count($bestCoveredSet)) {
                    $bestFound = $k;
                    $bestCoveredSet = $covered;
                }
            }
            // $needed -= $bestCoveredSet;
            $needed = array_diff($needed, $bestCoveredSet);
            $finalSet[] = $bestFound;
        }
        return $finalSet;
    }
}
